File: packages/provide-uterm/src/provide/uterm/control/plane/sqlite/connection.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
from __future__ import annotations

import re
import sqlite3
from pathlib import Path
from urllib.parse import unquote, urlparse

import aiosqlite

# sqlite:///C:/Users/... yields urlparse path "/C:/Users/..." — a leading "/"
# before a Windows drive letter, which Path() treats as a literal (invalid)
# folder segment rather than a drive-anchored path. The lookahead (rather than
# consuming the separator) also matches a bare drive with no trailing segment
# ("sqlite:///C:" -> path "/C:"), not just "/C:/...".
_WINDOWS_DRIVE_PATH_RE = re.compile(r"^/[A-Za-z]:(?=[/\\]|$)")

# sqlite://C:/Users/... (2 slashes, missing the 3rd) puts the drive letter in
# netloc instead of path ("C:", "/Users/...") -- a malformed-but-common typo
# of the 3-slash form, not a real host, so it's special-cased ahead of the
# general netloc handling below.
_DRIVE_LETTER_NETLOC_RE = re.compile(r"^[A-Za-z]:$")


class SqliteConnectionError(RuntimeError):
    """Raised when a SQLite control-plane connection cannot be initialized."""


def resolve_database_path(database_url: str) -> str:
    """Resolve a SQLite database URL or filesystem path to a connectable path."""
    if database_url in {":memory:", "file::memory:"}:
        return ":memory:"
    parsed = urlparse(database_url)
    if parsed.scheme in {"sqlite", "sqlite+aiosqlite"}:
        path = unquote(parsed.path or "")
        if path in {"", "/:memory:", ":memory:"}:
            return ":memory:"
        if _DRIVE_LETTER_NETLOC_RE.match(parsed.netloc):
            return f"{parsed.netloc}{path}"
        if _WINDOWS_DRIVE_PATH_RE.match(path):
            path = path[1:]
        if parsed.netloc:
            return f"//{parsed.netloc}{path}" if path.startswith("/") else f"//{parsed.netloc}/{path}"
        return path
    return database_url


async def connect_sqlite(database_url: str, *, busy_timeout_ms: int = 5_000, wal: bool = True) -> aiosqlite.Connection:
    """Open a SQLite connection with the baseline bootstrap pragmas applied."""
    database_path = resolve_database_path(database_url)
    if database_path != ":memory:":
        Path(database_path).expanduser().parent.mkdir(parents=True, exist_ok=True)
    conn = await aiosqlite.connect(database_path)
    try:
        conn.row_factory = sqlite3.Row
        await conn.execute(f"PRAGMA busy_timeout={int(busy_timeout_ms)}")
        if wal and database_path != ":memory:":
            await conn.execute("PRAGMA journal_mode=WAL")
        await conn.commit()
        return conn
    except Exception as exc:
        await conn.close()
        raise SqliteConnectionError(f"failed to initialize sqlite control-plane connection: {exc}") from exc

File: packages/provide-uterm-go/controlplane/sqlite/connection.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package sqlite

import (
	"context"
	"database/sql"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	_ "modernc.org/sqlite" // pure-Go SQLite driver; registers "sqlite".
)

// ConnectionError is returned when a SQLite control-plane connection cannot be
// initialized. Port of control.plane.sqlite.connection.SqliteConnectionError.
type ConnectionError struct{ msg string }

func (e *ConnectionError) Error() string { return e.msg }

// ResolveDatabasePath resolves a SQLite database URL or filesystem path to a
// connectable path. Port of control.plane.sqlite.connection.resolve_database_path.
func ResolveDatabasePath(databaseURL string) string {
	if databaseURL == ":memory:" || databaseURL == "file::memory:" {
		return ":memory:"
	}
	parsed, err := url.Parse(databaseURL)
	if err == nil && (parsed.Scheme == "sqlite" || parsed.Scheme == "sqlite+aiosqlite") {
		path, uerr := url.PathUnescape(parsed.Path)
		if uerr != nil {
			path = parsed.Path
		}
		if path == "" || path == "/:memory:" || path == ":memory:" {
			return ":memory:"
		}
		return path
	}
	return databaseURL
}

// conn is a live connection wrapper holding the single dedicated *sql.Conn that
// mirrors Python's one aiosqlite connection. All engine and store I/O runs on
// this conn; serialization is provided by the engine's tx-lock.
type conn struct {
	db   *sql.DB
	conn *sql.Conn
}

// connectSQLite opens a SQLite connection with the baseline bootstrap pragmas
// applied. Port of control.plane.sqlite.connection.connect_sqlite.
func connectSQLite(ctx context.Context, databaseURL string, busyTimeoutMS int, wal bool) (*conn, error) {
	dbPath := ResolveDatabasePath(databaseURL)
	if dbPath != ":memory:" {
		if dir := filepath.Dir(expandUser(dbPath)); dir != "" {
			if err := os.MkdirAll(dir, 0o755); err != nil {
				return nil, &ConnectionError{msg: fmt.Sprintf(
					"failed to initialize sqlite control-plane connection: %v", err)}
			}
		}
	}
	// MaxOpenConns(1) + a single grabbed *sql.Conn reproduces Python's single
	// aiosqlite connection: BEGIN IMMEDIATE and its COMMIT run on the same conn.
	db, err := sql.Open("sqlite", dbPath)
	if err != nil {
		return nil, &ConnectionError{msg: fmt.Sprintf(
			"failed to initialize sqlite control-plane connection: %v", err)}
	}
	db.SetMaxOpenConns(1)
	c, err := db.Conn(ctx)
	if err != nil {
		_ = db.Close()
		return nil, &ConnectionError{msg: fmt.Sprintf(
			"failed to initialize sqlite control-plane connection: %v", err)}
	}
	if _, err := c.ExecContext(ctx, fmt.Sprintf("PRAGMA busy_timeout=%d", busyTimeoutMS)); err != nil {
		_ = c.Close()
		_ = db.Close()
		return nil, &ConnectionError{msg: fmt.Sprintf(
			"failed to initialize sqlite control-plane connection: %v", err)}
	}
	if wal && dbPath != ":memory:" {
		if _, err := c.ExecContext(ctx, "PRAGMA journal_mode=WAL"); err != nil {
			_ = c.Close()
			_ = db.Close()
			return nil, &ConnectionError{msg: fmt.Sprintf(
				"failed to initialize sqlite control-plane connection: %v", err)}
		}
	}
	return &conn{db: db, conn: c}, nil
}

// close releases the dedicated conn and the pool.
func (c *conn) close() error {
	err := c.conn.Close()
	if dbErr := c.db.Close(); dbErr != nil && err == nil {
		err = dbErr
	}
	return err
}

// expandUser expands a leading ~ to the user's home directory, mirroring
// Python's Path.expanduser() used before mkdir.
func expandUser(p string) string {
	if p == "~" || strings.HasPrefix(p, "~/") {
		if home, err := os.UserHomeDir(); err == nil {
			if p == "~" {
				return home
			}
			return filepath.Join(home, p[2:])
		}
	}
	return p
}
